crates/future_metrics/src/lib.rs

```diff
@@ -0,0 +1,189 @@
+        Metered::new(self, labels)
+    }
+
+    /// A shortcut for [`FutureExt::with_labeled_metrics`] using a single label
+    /// only (presumably [`future_name`]).
+    fn with_metrics(self, label: &'static Label) -> Metered<Self> {
+        self.with_labeled_metrics(std::slice::from_ref(label))
+    }
+}
+
+impl<F> FutureExt for F where F: Future {}
+
+#[pin_project::pin_project]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct Metered<F> {
+    #[pin]
+    future: F,
+    state: State,
+}
+
+struct State {
+    started_at: Option<Instant>,
+    is_finished: bool,
+
+    poll_duration_total: Duration,
+
+    metrics: Metrics,
+}
+
+impl<F> Metered<F> {
+    fn new(future: F, metric_labels: &'static [Label]) -> Self {
+        let metrics = Metrics::new(metric_labels);
+
+        metrics.created.increment(1);
+
+        Self {
+            future,
+            state: State {
+                started_at: None,
+                is_finished: false,
+                poll_duration_total: Duration::from_secs(0),
+                metrics,
+            },
+        }
+    }
+}
+
+impl<F: Future> Future for Metered<F> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut this = self.project();
+        let s = &mut this.state;
+
+        if s.started_at.is_none() {
+            s.started_at = Some(Instant::now());
+            s.metrics.started.increment(1);
+        }
+
+        let poll_started_at = Instant::now();
+        let result = this.future.poll(cx);
+        let poll_duration = poll_started_at.elapsed();
+
+        s.metrics.poll_duration.record(poll_duration);
+        s.poll_duration_total += poll_duration;
+
+        if result.is_ready() && !s.is_finished {
+            s.is_finished = true;
+            s.metrics.finished.increment(1);
+            s.metrics
+                .duration
+                .record(s.started_at.unwrap_or(poll_started_at).elapsed())
+        }
+
+        result
+    }
+}
+
+impl Drop for State {
+    fn drop(&mut self) {
+        if !self.is_finished {
+            self.metrics.cancelled.increment(1);
+            if let Some(started_at) = self.started_at {
+                self.metrics.duration.record(started_at.elapsed())
+            }
+        }
+
+        self.metrics
+            .poll_duration_total
+            .record(duration_as_millis_f64(self.poll_duration_total));
+    }
+}
+
+struct Metrics {
+    duration: Histogram,
+
+    created: Counter,
+    started: Counter,
+    finished: Counter,
+    cancelled: Counter,
+
+    poll_duration: Histogram,
+    poll_duration_total: Histogram,
+}
+
+impl Metrics {
+    fn new(labels: &'static [Label]) -> Self {
+        metrics::with_recorder(|r| {
+            let metadata = Metadata::new(METADATA_TARGET, Level::INFO, None);
+
+            Self {
+                duration: r.register_histogram(
```

Comment:
Are you sure this references the same counters for all `Metrics` instances? With `opentelemetry` we had to use static/shared `Counter` instances, otherwise the exported data was incorrect.

Reply:
the same function is being called in their macro, so I'm relatively sure